extensions/ql-vscode/src/queries-panel/query-pack-discovery.ts

```diff
@@ -0,0 +1,93 @@
+import { basename, dirname } from "path";
+import { CodeQLCliServer, QuerySetup } from "../codeql-cli/cli";
+import { Event } from "vscode";
+import { App } from "../common/app";
+import { getOnDiskWorkspaceFolders } from "../helpers";
+import { dbSchemeToLanguage } from "../common/query-language";
+import { FALLBACK_QLPACK_FILENAME, QLPACK_FILENAMES } from "../pure/ql";
+import { FilePathDiscovery } from "./file-path-discovery";
+import { getErrorMessage, isEqualOrHigherPath } from "../pure/helpers-pure";
+import { extLogger } from "../common";
+import { EOL } from "os";
+
+export interface QueryPack {
+  path: string;
+  language: string | undefined;
+}
+
+/**
+ * Discovers all query packs contained in the workspace.
+ */
+export class QueryPackDiscovery extends FilePathDiscovery<QueryPack> {
+  constructor(app: App, private readonly cliServer: CodeQLCliServer) {
+    super("Query Pack Discovery", "**/{qlpack.yml,codeql-pack.yml}", app);
+  }
+
+  public get onDidChangeQueryPacks(): Event<void> {
+    return this.onDidChangePathsEmitter.event;
+  }
+
+  public findQueryPack(queryPath: string): QueryPack | undefined {
+    // Find all packs in a higher directory than the query
+    const packs = this.paths.filter((queryPack) =>
+      isEqualOrHigherPath(dirname(queryPack.path), queryPath),
+    );
+
+    // Sort by descreasing path length to find the pack nearest the query
+    packs.sort((a, b) => b.path.length - a.path.length);
+
+    // If the first two packs are from the same directory then look at the filenames
+    if (
+      packs.length >= 2 &&
+      dirname(packs[0].path) === dirname(packs[1].path)
+    ) {
+      if (basename(packs[0].path) === FALLBACK_QLPACK_FILENAME) {
```

Comment:
I think it does, though it may not be very clear. The `paths` array should be keeping track of every file with a name from `QLPACK_FILENAMES`, so this should include both kinds of pack names.

If there's both files present in a directory, what does CodeQL do? I've assumed that the one identified by `FALLBACK_QLPACK_FILENAME` (which is "qlpack.yml") wins and we use that file. That's what we're attempting to do `findQueryPack` by filtering to everything in a directory and then if we get two files trying to pick the right one.

Does that sound like the right logic?